### DISCUSSION TITLE
openai api compatibility

### DISCUSSION THREAD

**Discussion by iplayfast** [ggml-org/llama.cpp#3683] [General]
I know that llama-cpp-python has openai compatibility https://abetlen.github.io/llama-cpp-python/ but it seems to me this functionality would be better in llama.cpp's server. 

The server currently has an api, but I'm not aware of any documentation for it, or how compatible it is to openai.  
I'm using autogen with text-generation-webui and it works. I'd rather keep things simpler and just use the llama.cpp server.

**Comment by ianscrivener** [comment]
Agree!! It would be fantastic for llama.cpp to have an OpenAI compatible API.

**Reply by ianscrivener** [reply]
@iplayfast, 
Just had a play with llama-cpp-python server, the latest version. Works very well.

Other than being in C/C++... and likely running slightly faster... what other benefits do you see in C/C++ ?

What other features do you think would be good?

**Reply by iplayfast** [reply]
I expect the llama api was developed early on in the project when things were much more fluid. 

Currently llama.cpp has a server example with an api that isn't documented (as far as I can see). If the openai api is standardized on, it means one less thing to think about. That is, why have to api's when you can have one that is used universally. 

Another benefit that I'm looking for is to not have to run python to have a server running.

 With Llama.cpp server the memory footprint is smaller, and as I want this to be always running in the background while I work on other things, I want it as small as possible. I know it's not much of a difference, but why have two different api's.

**Reply by ianscrivener** [reply]
@iplayfast, fair enough...  and I agree with your rationale.

IMO having a fully featured and **self-documenting** llama.cpp API could be of huge benefit to the project... AND perhaps be the easest way to solve the current patchy documentation coverage. _Two birds with one stone..._ 

Swagger/OpenAPI is an **excellent** documentation/testing tool for app functionality - and with the right libraries can be **AUTO-DOCUMENTING**. I know that @ggerganov and the dev team are generally opposed to using external C/C++ libraries.. though there have been a couple of external libraries used in non-core examples.

Previously I built a very large API for an Australian company in node.js - with auto-documenting Swagger API. While there was slightly more effort to wire up the API and do the extra code for auto-documentation... the overall development effort was greatly reduced as there were far less questions and queries to the developer(s).

Oat++ [[https://github.com/oatpp/oatpp-swagger]] looks like a modern C++ helper library that deliver and API & Swagger/OpenAPI document and a Swagger API tester out of the box. 

I'd be keen to here what @ggerganov and the main C/C++ devs think of the idea...

**Reply by ianscrivener** [reply]
@ggerganov, could running python or curl tests against a comprehensive API be helpful for testing llama.cpp in CI/CD? I've used this approach/architecture in large node.js API apps with success as part of a CI/CD testing suite.

PS: while I'm not a C/C++ dev... I'd be very happy to do the scaffolding and boiler plate work on the API and contribute as best I can. I'm fairly familiar with APIs, Swagger and building API products etc.

**Reply by shibe2** [reply]
Current server API is lower level and allows building different APIs on top of it. One thing that middleware usually has to do is prompt formatting. Prompt format depends on the model used. New fine-tunes may introduce new formats or variations of existing formats. Maybe it's best to keep this stuff separate for the time being and avoid endless feature requests for new formats. Hugging Face has [chat templates](https://huggingface.co/docs/transformers/main/chat_templating) using Jinja. One could work on adding it to llama.cpp, initially in Python.

**Reply by ianscrivener** [reply]
**auto-generating**, **auto-documenting** APIs are straight forward and flexible. I'd suggest both a full API equivalent to llama.cpp ./main as well as OpenAI compatible endppoints - eg;
```
/openai/v1/chat/completions
/openai/v1/completions
/openai/v1/embeddings
/llama-cpp/v1/xxxx
```

**Reply by ianscrivener** [reply]
Yes, good point re prompt formatting
Prompt formats are data - and should be separate from code, like models. LM-Studio has handled [models](https://github.com/lmstudio-ai/model-catalog/tree/main/models) and [inference config (including prompts)](https://gist.github.com/ianscrivener/9c57078a6d2e8891f83c26fc34e19d4f) very elegantly
 
Default and custom prompt formats would be easy to do with a well designed API 

For an API I'd be suggesting full llama.cpp funcitonality.. in addition to OpenAI compatible endpoints for LLM inference and embeddings (not all OpenAI stuff). Eg;
```
/llama_cpp/xxxx .. _essentially ./main_ 
/llama_cpp/prompts/
/llama_cpp/prompts/{prompt}
/llama_cpp/llava
/llama_cpp/models
/llama_cpp/models/{model}
/llama_cpp/perplexity

openai/chat/completions
openai/completions
openai/embeddings
openai/models
openai/models/{model}
```

**Reply by shibe2** [reply]
>Lm-Studio has handled models and prompts very elegantly - https://github.com/lmstudio-ai/model-catalog/tree/main/models

I'm not familiar with it, and I don't see anything related to prompts at that URL.

**Reply by ianscrivener** [reply]
Sorry, here's an exampe of LM Studio inference config including prompt: 
https://gist.github.com/ianscrivener/9c57078a6d2e8891f83c26fc34e19d4f
(I've updated my post)

**Reply by shibe2** [reply]
Does it handle special tokens like in ChatML?

**Reply by ianscrivener** [reply]
```
/llama_cpp/prompts                       ... baked in prompts & custom prompts
/llama_cpp/prompts/{prompt}
/llama_cpp/presets                        ... inference param presets, including referece to prompt
/llama_cpp/presets/{preset}
```

**Comment by shibe2** [comment]
Have you tried [api_like_OAI.py](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/api_like_OAI.py)?

**Comment by shibe2** [comment]
[Server API documentation](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md#api-endpoints). It's not comprehensive and may get out of date sometimes, but it's enough to get you started.

**Reply by theoparis** [reply]
This doesn't seem to have functions/tools support 😕

**Comment by iplayfast** [comment]
@shibe2 thanks for those links. Was not aware of either of them.
